### PR TITLE
[0.24.0] Authentication: Azure Active Directory

### PR DESCRIPTION
**Pull request by jeliasson**
Updating docs for Azure Active Directory authentication to be released with `0.24.0`.

#### Related PR
https://github.com/redwoodjs/redwood/pull/1675

#### Related Issue
[redwoodjs/redwood#1672](https://github.com/redwoodjs/redwood/issues/1672#issuecomment-769766607)

### CODE REVIEW

File: docs/authentication.md [hunk 1]
```diff
@@ -290,9 +285,32 @@ ReactDOM.render(
 
 To setup your App Registration with custom roles and have them exposed via the `roles` claim, follow [this documentation](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-add-app-roles-in-azure-ad-apps).
 
-#### Login and Logout Options
+#### Login Options
+
+Options in method `logIn(options?)` is of type [AuthRequest](https://pub.dev/documentation/msal_js/latest/msal_js/AuthRequest-class.html) and is a good place to pass in optional [scopes](https://docs.microsoft.com/en-us/graph/permissions-reference#user-permissions) to be authorized. By default, MSAL set `scopes` to [/.default](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#the-default-scope) which is built in for every application that refers to the static list of permissions configured on the application registration.
```
Comment (Tobbe) [inline docs/authentication.md:290]:
```suggestion
Options in method `logIn(options?)` is of type [AuthRequest](https://pub.dev/documentation/msal_js/latest/msal_js/AuthRequest-class.html) and is a good place to pass in optional [scopes](https://docs.microsoft.com/en-us/graph/permissions-reference#user-permissions) to be authorized. By default, MSAL sets `scopes` to [/.default](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#the-default-scope) which is built in for every application that refers to the static list of permissions configured on the application registration.
```

Can some native English speaker please check on this. Should it be "set" or "sets"?

File: docs/authentication.md [hunk 2]
```diff
@@ -290,9 +285,32 @@ ReactDOM.render(
...
+  scopes: ['User.Read.All'],
+})
+```
+
+See [loginPopup](https://pub.dev/documentation/msal_js/latest/msal_js/UserAgentApplication/loginPopup.html) or [full class documentation](https://pub.dev/documentation/msal_js/latest/msal_js/UserAgentApplication-class.html#constructors) for more information.
+
+#### getToken Options
+
+Options in method `getToken(options?)` is of type [AuthRequest](https://pub.dev/documentation/msal_js/latest/msal_js/AuthRequest-class.html).
+By default, method `getToken` will be called with options below. Essentially this is MSAL's way of using the same scope that was used in the `logIn` request.
```
Comment (Tobbe) [inline docs/authentication.md:303]:
```suggestion
By default, method `getToken` will be called with the options below. Essentially this is MSAL's way of using the same scope that was used in the `logIn` request.
```

File: docs/authentication.md [hunk 3]
```diff
@@ -262,7 +258,6 @@ The Authority is a URL that indicates a directory that MSAL can request tokens f
 
 ```js
 // web/src/index.js
-import { AuthProvider } from '@redwoodjs/auth'
```
Comment (Tobbe) [inline docs/authentication.md:265]: Why did you remove this?

File: docs/authentication.md [hunk 4]
```diff
@@ -290,9 +285,32 @@ ReactDOM.render(
...
+See [loginPopup](https://pub.dev/documentation/msal_js/latest/msal_js/UserAgentApplication/loginPopup.html) or [full class documentation](https://pub.dev/documentation/msal_js/latest/msal_js/UserAgentApplication-class.html#constructors) for more information.
+
+#### getToken Options
+
+Options in method `getToken(options?)` is of type [AuthRequest](https://pub.dev/documentation/msal_js/latest/msal_js/AuthRequest-class.html).
+By default, method `getToken` will be called with options below. Essentially this is MSAL's way of using the same scope that was used in the `logIn` request.
+
+```js
+await getToken({
+  scopes: [process.env.AZURE_ACTIVE_DIRECTORY_CLIENT_ID],
```
Comment (Tobbe) [inline docs/authentication.md:307]: Should the client id really be passed as a scope?

Comment (jeliasson) [reply docs/authentication.md:307]: Hmm, I think I scrolled down in the document and realized that MagicLink and Firebase do not reference these in the docs (I guess for brevity, the same as `FatalErrorBoundary` and `Redwood(Apollo)Provider`). However, when reading your comment I realize that it's there on the remaining providers.

I'll re add it for now. 😊

File: docs/authentication.md (hunk 3 above)
Comment (jeliasson) [reply docs/authentication.md:265]: Hmm, I think I scrolled down in the document and realized that MagicLink and Firebase do not reference these in the docs (I guess for brevity, the same as FatalErrorBoundary and Redwood(Apollo)Provider). However, when reading your comment I realize that it's there on the remaining providers.

I'll re add it for now. 😊

File: docs/authentication.md (hunk 4 above)
Comment (jeliasson) [reply docs/authentication.md:307]: From my understanding, passing the client id will translate into default scopes of `['openid', 'profile']` or re-use the scopes that was used in the login methods (`loginPopup` or `loginRedirect`) without being explicit about it. Unfortunately the documentation about this is very slim and the majority of their examples explicitly write out the same scope as their login scope(s).

We can consider hard-coding `['openid', 'profile']` to make this more tangible, but run the risk of missing out on future defaults without updating, or hard-code `/.defaults` but this would scope _all_ permissions on that app.

**Special OIDC Scopes behavior cases**
https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-core/docs/scopes.md#special-oidc-scopes-behavior-cases

**Related**
https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/2536

Let me know your thoughts.

File: docs/authentication.md (hunk 2 above)
Comment (jeliasson) [reply docs/authentication.md]: Thanks for spotting that!

File: docs/authentication.md (hunk 1 above)
Comment (jeliasson) [reply docs/authentication.md]: I'm pretty sure that it is _sets_ as you suggested, but let's keep this open for a bit.

File: docs/authentication.md (hunk 4 above)
Comment (dthyresson) [reply docs/authentication.md:308]: I read the Special OIDC behavior above and it says that the

> ClientId is removed from the scopes array when it is the only scope in the configuration. If it is not the only scope, it is treated as a resource scope and will be sent in the final server request.

So, it is important that it be the *only* scope 

>    { scopes: ['YOUR_CLIENT_ID'] } // becomes { scopes: ['openid', 'profile'] } before the request is sent (ClientId is spliced out)

otherwise:

>    { scopes: ['YOUR_CLIENT_ID', 'User.Read'] } // becomes { scopes ['YOUR_CLIENT_ID', 'User.Read', 'openid', 'profile']} before the request is sent (ClientId is treated as resource scope and therefore not spliced out)

>    { scopes: ['YOUR_CLIENT_ID', 'openid'] } // becomes { scopes ['YOUR_CLIENT_ID', 'openid', 'profile']} before the request is sent

So it should be fine, but perhaps a comment noting if any other scopes are included on that line 308 it will also be sent (if that's something one would want, I do not know).

Comment (Tobbe) [reply docs/authentication.md:308]: > ClientId is removed from the scopes array when it is the only scope in the configuration.

To me that sounds as though we could just as well only have this (assuming `{ scopes: [] }` is the default)
```js
await getToken()
``` 

Am I right? And, if so, do we want to simplify the code, or do we want to keep it explicit?

Comment (dthyresson) [reply docs/authentication.md:308]: @Tobbe You might be right b/c the docs say:

> For this reason, whether or not the developer adds the openid or profile scopes to their request configuration, msal@1.x will make sure they are included before sending the request to the STS.

That is, `'openid', 'profile'` will *always* be included as scopes.

Therefore, 

`await getToken()`

seems like it would be sufficient.

Comment (jeliasson) [reply docs/authentication.md:308]: You guys are right. I believe I saw a comment that the `client_id` will use `['openid', 'profile']` _or_ the scopes that was defined in the login request. However, I can't find this reference anymore.

> For this reason, whether or not the developer adds the openid or profile scopes to their request configuration, [msal@1.x](mailto:msal@1.x) will make sure they are included before sending the request to the STS.

As this is the case, we don't need to scope to `['<client_id>']` as default but rather `[]`. 
We would still want to have `options?` to allow overriding this, right? Otherwise I'm not sure how one would extend the scope if not for below.
```js
await getToken({
  scopes: ['User.ReadWrite.All'] // becomes ['User.ReadWrite.All', 'openid', 'profile']
})
```

I'll make some tests and get back to you. In the meantime, feel free to elaborate on how the docs should be written.

Comment (jeliasson) [reply docs/authentication.md:308]: > You guys are right. I believe I saw a comment that the `client_id` will use `['openid', 'profile']` _or_ the scopes that was defined in the login request. However, I can't find this reference anymore.

They apply incremental consent as described below and thus confirms that we don't need to pass in client id as default. 
https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/resources-and-scopes.md#dynamic-scopes-and-incremental-consent

Comment (jeliasson) [reply docs/authentication.md]: > 
>
> > ClientId is removed from the scopes array when it is the only scope in the configuration.
> 
> To me that sounds as though we could just as well only have this (assuming `{ scopes: [] }` is the default)
> 

@Tobbe @dthyresson 
Hmm... passing/default to a empty scopes (i.e. `{ scopes: [] }`) does not work. No errors, no sign in... 🤔

File: docs/authentication.md (hunk 1 above)
Comment (dthyresson) [reply docs/authentication.md]: :+1: sets

File: docs/authentication.md (hunk 4 above)
Comment (Tobbe) [reply docs/authentication.md]: > Hmm... passing/default to a empty scopes (i.e. { scopes: [] }) does not work. No errors, no sign in... 🤔
What do we say of having openid and profile as defaults?

What?! That's like totally contradictory to their docs, no? What if you pass an empty object, or nothing at all?

Comment (jeliasson) [reply docs/authentication.md]: > What?! That's like totally contradictory to their docs, no? 

I have not seen any documentation on wherever `scopes` can be null or empty array, and it does not look like it. 

> What if you pass an empty object, or nothing at all?

No, no luck. I have tested the following;

✅ **client_id**
```js
const authRequest = options || {
  scopes: [process.env.AZURE_ACTIVE_DIRECTORY_CLIENT_ID], 
}
```

✅ **default openid and profile**
```js
const authRequest = options || {
  scopes: ['openid', 'profile'], 
}
```

❌ **empty array**
```js
const authRequest = options || {
  scopes: [], 
}
```

❌ **null scopes**
```js
const authRequest = options || {
  scopes: null, 
}
```

❌ **empty object**
```js
const authRequest = options || {}
```

❌ **null**
```js
const authRequest = options || null
```

I think that defaulting to `['openid', 'profile']` is the way to go here to be honest.

Comment (Tobbe) [reply docs/authentication.md]: > I think that defaulting to ['openid', 'profile'] is the way to go here to be honest.

👍 Makes it very clear exactly what scopes are used

**Comment by jeliasson** [comment]
@dthyresson @Tobbe 
Please review https://github.com/redwoodjs/redwood/pull/1703 and [updated docs](https://deploy-preview-534--redwoodjs.netlify.app/docs/authentication#azure-active-directory) at your convenience. 

Also, I would like to squash all my commits in this feature branch into one. Any git-ninja that knows how to do this? 

/cc @thedavidprice

**Comment by Tobbe** [comment]
> Also, I would like to squash all my commits in this feature branch into one. Any git-ninja that knows how to do this?

I think we can do that when we merge
image

Otherwise you can do it locally with something like `git rebase -i` and then force pushing, `git push -f`

**Comment by thedavidprice** [comment]
All our repos have a default setting to "Squash and Merge". All good!
Screen Shot 2021-02-01 at 3 21 46 PM

Thanks, gents!
